### ISSUE TITLE
Make `@` a special form to allow easier replacements

### ISSUE THREAD

**Issue by josevalim** [elixir-lang/elixir#846]
One thing that is not easy in Elixir is to dynamically generate a function based on compile time values. For example, imagine you have this keyword arguments:

```
[foo: 1, bar: :baz, bat: :zoo]
```

And you want to generate a bunch of functions with names equal to key and that return their respective values. In Elixir, you have a couple approaches to solve this. You could create your own macro, you could eval some code or you can use `def/4`. The simplest approach in my opinion is `def/4`:

```
Enum.each list, fn { key, value } ->
  def key, [], [], do: Macro.escape(value)
end
```

However, the example above is too simple. Let's consider a more complex one. Let's suppose our key/values represent ansi escape codes and we want to generate two functions for each sequence, similar to [IO.ANSI](https://github.com/elixir-lang/elixir/blob/19d402b7d5280fb54a213f0a699a7daed54a9030/lib/elixir/lib/io/ansi.ex#L5). Our functions would then need to be defined as follows:

```
Enum.each [bright: 1, faint: 2], fn { name, code } ->
  def name, [], [], do:
    quote do: "\e[#{unquote(code)}m"

  args =
    quote do: [<< unquote(atom_to_binary(name)), rest :: binary >> ]

  defp :escape_sequence, args, [], do:
    quote do: { "\e[#{unquote(code)}m", rest }
end
```

The example above gets considerably harder and more verbose. This pull request proposes the following to be possible:

```
Enum.each [bright: 1, faint: 2], fn { name, code } ->
  @name unquote(name)
  @code unquote(code)
  @bin  atom_to_binary(name)

  def @name, do: "\e[#{@code}m"

  defp escape_sequence(<< @bin, rest :: binary >>), do:
    { "\e[#{@code}m", rest }
end
```

The second example is considerably less verbose.

Notice that [module attributes are already used as value replacement](https://github.com/elixir-lang/elixir/blob/master/lib/elixir/lib/hash_dict.ex#L48), serving as what other languages would call constants. This pull request is only about making the following line work:

```
  def @name, do: "\e[#{@code}m"
```

Today, the line above is parsed as:

```
   def @(name), do: "\e[#{@code}m"
```

So it is effectively defining a function named `@` which receives `name` as argument. By making `@` a special form, it cannot be defined by users, so we can effectively translate `@name` to the current value of `@name`.

Summing up: this proposes intendets to make definition of dynamic functions easier by extending `@` to a special form. The downside is that developers will no longer be able to redefine `@` to their own semantics, but it gives more flexibility to every day usage of Elixir.
